Generate subject for email preview and test email

NeosinneR opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Part of #52228. This PR uses the backend to generate an email subject for email preview and when sending a test email. It also adds a new `woocommerce_email_preview_placeholders` filter.

Closes #53575.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Subject line**
1. Enable `email_improvements` experimental flag (in DB, set `woocommerce_feature_email_improvements_enabled` option to `yes`).
2. Go to **WooCommerce > Settings > Emails**.
3. Verify that the email preview has a subject line.
4. Change the email type in the email preview and verify that the subject line has changed.
5. Verify that placeholders (such as `{order_number}` or `{site_title}`) are replaced with actual values.
6. If your test site can send emails, send a test email and check that the subject line is also correct. 

**New filter**
1. Enable `email_improvements` experimental flag (in DB, set `woocommerce_feature_email_improvements_enabled` option to `yes`).
2. Go to **WooCommerce > Settings > Emails**.
3. Manage any email and add a new custom placeholder to a subject line (e.g., `{my_custom_placeholder}`).
4. Add a new filter to replace your placeholder (e.g., into your `functions.php` file):
```php
add_filter( 'woocommerce_email_preview_placeholders', function ( $placeholders ) {
	$placeholders['{my_custom_placeholder}'] = 'Hello, world!';
	return $placeholders;
}, 10, 1 );
```
5. Check that the placeholder is replaced in the email preview. 

<!-- End testing instructions -->

NeosinneR committed:
Move email preview setup to earlier step (from render to setting email type)

NeosinneR committed:
Set email type unless explicitly set before

NeosinneR committed:
Set up order-related email placeholders

NeosinneR committed:
Set proper subject when sending a test email

NeosinneR committed:
Move getting EmailPreview instance into constructor

NeosinneR committed:
Add GET route to get email subject for preview

NeosinneR committed:
Fetch email preview subject from API

NeosinneR committed:
Add changelog

NeosinneR committed:
Add REST API tests for preview subject

NeosinneR committed:
Add unit tests for EmailPreview

alex-mpoet reviewed (commented):
@NeosinneR Works as specified, but I have a few notes:
1. There's a minor visual issue. The height of the top bar and the name/email element position in the preview window are different before and after displaying the subject. It would be neat to maintain consistent height:
![image](https://github.com/user-attachments/assets/271841cd-a33e-41b3-82cf-a3ef295ccb5a)
2. We now have a special filter for preview placeholders (`woocommerce_email_preview_placeholders`), but what's the purpose of having placeholders that only work in a preview? Will they work in real emails at some point? Also, from a third-party developer standpoint, it might be more intuitive to have only one filter that adds placeholders for both preview and real emails, rather than having two separate filters. What do you think?
3. Typo in [commit message](https://github.com/woocommerce/woocommerce/pull/53581/commits/cc36d011d69465c1768e377e990b5b20eada4168): explicitely => explicitly.

NeosinneR committed:
Remove no longer needed subject in EmailType

NeosinneR committed:
Add zero-width space to prevent jumping when subject is async loaded

NeosinneR committed:
Add filter for email preview placeholders

NeosinneR commented:
@alex-mpoet
1. Nice catch. I haven't noticed it because the API request is fairly quick in my local environment. Addressed in b3c3bfc.
2. I don't think this is possible with how emails are handled in WooCommerce. For example, WooCommerce Subscriptions:
    - In some emails, they use `{customers_first_name}` or `{time_until_renewal}`. 
    - Even in the core, placeholders are [defined in `__construct()` as empty](https://github.com/woocommerce/woocommerce/blob/2aacfd90830a099cc08c8fd646881b883042648b/plugins/woocommerce/includes/emails/class-wc-email-cancelled-order.php#L35-L39), and only in the `trigger()` method are they [replaced with actual values](https://github.com/woocommerce/woocommerce/blob/2aacfd90830a099cc08c8fd646881b883042648b/plugins/woocommerce/includes/emails/class-wc-email-cancelled-order.php#L87-L89). 
    - The same is with extensions - [defined in `__construct()`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/1692c41cf43e2c3c5ef0816a48196ffb241b9e72/includes/emails/class-wcs-email-customer-notification.php#L16-L22) but [replaced in `trigger()`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/1692c41cf43e2c3c5ef0816a48196ffb241b9e72/includes/emails/class-wcs-email-customer-notification.php#L94-L95). 
    - But `trigger()` can't be used in email preview because that method also sends the email, which we don't want to do.
    - Without this filter, it wouldn't show any error, but the placeholders would be replaced with an empty string. E.g. here it's missing customer's name. 
        <img width="705" alt="Screenshot 2024-12-10 at 22 25 35" src="https://github.com/user-attachments/assets/7e48a03a-161c-4eb9-9630-210d11e3eda5">
    - The reason I added the filter is to provide a way for extensions to fix that if they want to. Yes, it'll duplicate the same definition that's already in `trigger()` methods, but considering the scope of this project, I don't think there is an easy way to solve this without duplication. What do you think? 
3. Fixed

alex-mpoet reviewed (approved):
> Yes, it'll duplicate the same definition that's already in trigger() methods, but considering the scope of this project, I don't think there is an easy way to solve this without duplication. What do you think?

Thanks for the explanation! I understand now it's a workaround that's only needed for previews. I agree that eliminating duplication is out of scope of the project.

Everything looks good! Approved!
